sql/src/main/java/io/crate/expression/reference/sys/shard/ShardSegments.java

```diff
@@ -0,0 +1,71 @@
+/*
+ * Licensed to Crate under one or more contributor license agreements.
+ * See the NOTICE file distributed with this work for additional
+ * information regarding copyright ownership.  Crate licenses this file
+ * to you under the Apache License, Version 2.0 (the "License"); you may
+ * not use this file except in compliance with the License.  You may
+ * obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or
+ * implied.  See the License for the specific language governing
+ * permissions and limitations under the License.
+ *
+ * However, if you have executed another commercial license agreement
+ * with Crate these terms will supersede the license and you may use the
+ * software solely pursuant to the terms of the relevant commercial
+ * agreement.
+ */
+
+package io.crate.expression.reference.sys.shard;
+
+import org.elasticsearch.cluster.routing.ShardRouting;
+import org.elasticsearch.cluster.service.ClusterService;
+import org.elasticsearch.common.inject.Inject;
+import org.elasticsearch.common.inject.Singleton;
+import org.elasticsearch.index.IndexService;
+import org.elasticsearch.index.shard.ShardId;
+import org.elasticsearch.indices.IndicesService;
+
+import java.util.Iterator;
+import java.util.stream.Stream;
+
+@Singleton
+public class ShardSegments implements Iterable<ShardSegment> {
+
+    private final ClusterService clusterService;
+    private final IndicesService indicesService;
+
+    @Inject
+    public ShardSegments(ClusterService clusterService, IndicesService indicesService) {
+        this.clusterService = clusterService;
+        this.indicesService = indicesService;
+    }
+
+    @Override
+    public Iterator<ShardSegment> iterator() {
+        return clusterService
+            .state()
+            .getRoutingTable()
+            .allShards()
+            .stream()
+            .flatMap(this::buildShardSegment)
+            .iterator();
+    }
+
+    private Stream<ShardSegment> buildShardSegment(ShardRouting shardRouting) {
```

Comment:
Using the routing table/shard routing to get the shard ids will result in shard ids which may not available on each node. Thus this may result in ShardNotFoundException. `IndicesService.iterator()` should be used instead to get the local `IndexService` instances and `IndexService.iterator()` to get the local `IndexShard` instances.